### ISSUE TITLE
reactions: Label text can linebreak sooner than is strictly necessary

### ISSUE THREAD

**Issue by chrisbobbe** [zulip/zulip-flutter#433]
In a reaction chip in the message list (#410), the label with users' names (seen with `UserSettings.displayEmojiReactionUsers`) can get linebreaked for text wrapping sooner than is strictly necessary:

<img width="449" alt="image" src="https://github.com/zulip/zulip-flutter/assets/22248748/eb651e8f-52d5-4750-906d-57b0399d06fe">

In this example, it says:

> Cordelia, Lear's daughter, You,
> Polonius

when there is actually enough room to say

> Cordelia, Lear's daughter, You, Polonius

This is a tradeoff we made consciously when implementing a reasonable layout for text emoji, when both the emoji text and the reaction label are long. See the first two pairs of screenshots at https://github.com/zulip/zulip-flutter/pull/410#issuecomment-1833378268, which highlight the upside of that tradeoff. If we were less careful about text emoji, they could look extremely broken.

If we decide that the current issue is worth fixing, we should try to fix it without breaking text emoji. And we should keep in mind that `UserSettings.emojiset` isn't enough to determine whether we're going to show a text emoji, because we show text emoji for various error conditions, including when an image emoji fails to load, which we learn about especially late.

**Comment by CompleteGlory** [comment]
**I noticed this issue isn't tagged "help wanted" — is it open for someone to pick up? I'd like to work on it if so.**

I looked into it and the fix seems pretty small: in ReactionChip widget, the label width always gets narrowed to 75%, even for normal emoji where it's not needed. Only text emoji actually need that. 
There's already a TextEmojiDisplay type resolved a few lines up that I can check against.

There's one part I can't fully fix here ,when an image emoji fails to load and falls back to text, that happens later inside EmojiWidget's error builder, after this width is already decided. I'd leave that as is unless you want it handled too.

Let me know if this sounds right and if I can go ahead.

**Comment by chrisbobbe** [comment]
> when an image emoji fails to load and falls back to text

Right; this case was highlighted in the issue description. What does it look like in this case when the emoji label and the list of names are both very long? Would you post a screenshot?

**Comment by CompleteGlory** [comment]
### Here's what this looks like on a real device with a normal emoji vs a text emoji.

Unicode/image emoji, wraps to 2 lines vs fits on 1:
<img width="260" height="578" alt="Image" src="https://github.com/user-attachments/assets/7127502e-a845-4f1d-bf84-2aad871e3595" /> <img width="260" height="578" alt="Image" src="https://github.com/user-attachments/assets/6bbf8173-2161-4ccc-af42-dbbb66dbaceb" />

**Text emoji:**
<img width="260" height="578" alt="Image" src="https://github.com/user-attachments/assets/213f141e-2b26-48ab-92ce-aaa661d2d60a" /> <img width="260" height="578" alt="Image" src="https://github.com/user-attachments/assets/3a64fb1e-8788-4129-b348-77745983c4a4" />

**Comment by chrisbobbe** [comment]
I'm confused; what's meant to change between the first and second screenshot under "Unicode/image emoji"? And under "Text emoji"? Those don't look like "before" and "after" pairs, because the emoji chip stretches full-width in both.

**Comment by CompleteGlory** [comment]
These are not before and after, I was testing single line versus multiline behavior.
However in all cases that's after the changes made, if you need me to provide a before and after comparisons I can do so.

**Comment by chrisbobbe** [comment]
I see; thanks. I want to see this part:

> There's one part I can't fully fix here

I suspect your "Text emoji" screenshots don't actually show what's broken. Did you take those screenshots by simulating a failed image load, or by switching to the text-emoji setting?

**Comment by CompleteGlory** [comment]
No, I switched to text-emoji setting.
I'll try to simulate the failure and provide a screenshot as well.

**Comment by CompleteGlory** [comment]
I tried to simulate the failure manually and here's the long-label + long-name case:

<img width="270" alt="Image" src="https://github.com/user-attachments/assets/7ed8fa05-d6a9-409b-a276-ff0cce2fff74" /> 

<img width="270" alt="Image" src="https://github.com/user-attachments/assets/557b97c0-03a3-4c14-bdf1-4dd031fc042c" />

<img width="270" alt="Image" src="https://github.com/user-attachments/assets/76ea4975-b705-48e3-b08e-bb327f52753f" />

Realistically an emoji name probably wouldn't get this long, but it's the biggest edge case I could put together to stress-test the layout.

**Comment by chrisbobbe** [comment]
> Realistically an emoji name probably wouldn't get this long

Your example looks like the same length as `:family_woman_woman_girl_girl:`, which is a real emoji name. And I'm not sure what the cap is on custom emoji names; I think we should treat your example name as realistic.

In the middle screenshot, the button looks quite a bit taller than it needs to be: wouldn't it fit better if the emoji name and label text were both like 2 lines tall instead of 6?

**Comment by CompleteGlory** [comment]
#### I Think I've now handled that case and organized my changes a bit
when an image emoji fails to load, the chip rebuilds and switches to the same layout used for text emoji (label capped at 75% of the row). So the failed-image case now renders exactly like the text-emoji-setting case, with the emoji name and label each wrapping to about 2 lines instead of one tall column.

Implementation-wise: ReactionChip is stateful, and EmojiWidget gained an onImageError callback that fires from the image error builder; the chip schedules a post-frame rebuild when that happens (and resets if the chip gets reused for a different emoji).

Here's the same long-label + long-name case after the change, with the image failure simulated in 1 , 2 and 3 lines cases:

<img width="260"  alt="Image" src="https://github.com/user-attachments/assets/88be485e-10fb-4e28-a5c8-449567eb5e70" />
<img width="260"   alt="Image" src="https://github.com/user-attachments/assets/24966734-051f-4b63-a8c9-87092105ee41" />
<img width="260"  alt="Image" src="https://github.com/user-attachments/assets/20d7f3d1-45d3-4f97-90b1-25fe3fb024cc" />
